### PR TITLE
better direction calculation

### PR DESCRIPTION
**Pull request by pionere**
- no limitation on the vector's length
- no need for error-prone tables

### CODE REVIEW

**Comment by glebm** [comment]
I wonder if this function is a hotspot? The table version has fewer branches.

**Comment by pionere** [comment]
> I wonder if this function is a hotspot? The table version has fewer branches.

what are you talking about?
table version: 6 branches in any case
new version: 4 branches at most

**Comment by glebm** [comment]
I was estimating the number of branches by looking at the level of nesting. After looking more carefully, I see that it shouldn't be an issue. Perhaps this improves performance even?

**Comment by AJenbo** [comment]
Wrote a test, there is a slight difference in how `GetDirection16` works now:
```
SourceT/missiles_test.cpp:50: Failure
Expected equality of these values:
  1
  dvl::GetDirection16(0, 0, 8, 11)
    Which is: 0
SourceT/missiles_test.cpp:58: Failure
Expected equality of these values:
  9
  dvl::GetDirection16(8, 11, 0, 0)
    Which is: 8
SourceT/missiles_test.cpp:66: Failure
Expected equality of these values:
  13
  dvl::GetDirection16(0, 8, 11, 0)
    Which is: 12
```

**Comment by AJenbo** [comment]
It looks like the main cardinal directions become narrower with this change:
dir8

File: Source/missiles.cpp [hunk 1]
```diff
@@ -261,113 +261,130 @@ int GetSpellLevel(int id, int sn)
...
+			md = 2; // DIR_W
+		} else {
+			my = -my;
+			if (5 * mx <= (my << 1))
+				return 5; // DIR_NE
+			md = 4; // DIR_N
+		}
+		if (5 * my <= (mx << 1))
+			md = 3; // DIR_NW
+	}
```
Comment (julealgon) [inline Source/missiles.cpp:314]: Seems to me like the `if` block and the `else` block are pretty much identical. Is there no way to share the logic here?

I see the original code was using `abs`. Can't we do something similar to avoid having to duplicate all the checks?

**Review by julealgon — COMMENTED** [review]
This code has a very considerable amount of logic duplicated and is incredibly hard to follow (at least for me).

**Comment by galaxyhaxz** [comment]
Personally I'm in favor of it, with a bit of tweaking to get it closer to vanilla. Replacing hardcoded tables and other crap is a good thing, and I prefer the code method!

**Comment by AJenbo** [comment]
After some investigation we have concluded that the original isn't very accurate:
image

The new one pretty much fixes this issue:
image

In terms of gameplay impact, this only consequence seems to be that charging monsters will now be slightly more accurate when moving in some directions.

**Review by AJenbo — APPROVED** [review]

**Comment by AJenbo** [comment]
Merged via https://github.com/diasurgical/devilutionX/commit/79ed3c897bb0abd03024c5433ed47939bd738d06

File: Source/missiles.cpp (hunk 1 above)
Comment (AJenbo) [reply Source/missiles.cpp:314]: @pionere did clean it up, but never updated the PR so some one will  have to go dig it out:
https://github.com/pionere/devilutionX/blob/master/Source/missiles.cpp#L285

Comment (AJenbo) [reply Source/missiles.cpp:314]: @pionere has made a more minimalistic version, but some one would have to go dig it out:
https://github.com/pionere/devilutionX/blob/master/Source/missiles.cpp#L285
